File: leetcode/HashMap/0219_Contains_Duplicate_II.py
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        N = len(nums)
        index_dict = {}
        for i, num in enumerate(nums):
            if num not in index_dict:
                index_dict[num] = i
            else:
                if i - index_dict[num] <= k:
                    return True
                else:
                    index_dict[num] = i

        return False

File: leetcode/HashMap/0219_Contains_Duplicate_II.cpp
// O(N), use map to record index
class Solution {
public:
	bool containsNearbyDuplicate(vector<int>& nums, int k) {
		if (k <= 0)
			return false;

		map<int, int> pos;
		for(int i = 0; i < nums.size(); i++) {
			if (pos.find(nums[i]) != pos.end() && i - pos[nums[i]] <= k)
				// appeared, compare the distance
				return true;
			else
				// add new entry in map, or update the index
				pos[nums[i]] = i;
		}
		return false;
	}
};

// O(N), use set to record window with size k
class Solution {
public:
	bool containsNearbyDuplicate(vector<int>& nums, int k) {
		unordered_set<int> set;
		for (int i = 0; i < nums.size(); i++) {
			if (set.find(nums[i]) != set.end())
				return true;

			set.insert(nums[i]);
			if (i >= k)
				set.erase(nums[i - k]);
		}
		return false;
	}
};